### PR TITLE
Add section on managing abstraction v complexity

### PR DESCRIPTION
**Pull request by jasonyavorsky**
After reading https://future.a16z.com/the-case-for-developer-experience/, I felt like there was something interesting to say about our own product and how we think about the Code Graph in particular. I tried to apply what I read in the article to what I understand of the product today, but probably missed the mark somewhere or other.

I'm curious what everyone thinks of something like this. Does it resonate? If so, why or why not? Can we refine it and merge? Is it a bad idea? I'd love to hear everyone's perspectives.

### CODE REVIEW

**Comment by jasonyavorsky** [comment]
@malomarrec @macraig @Joelkw @benvenker @lguychard as PMs for this product area, I don't want to merge this without your affirmation. Please take a look when you can and approve the PR if you're good with it, or comment with why not so we can discuss. Thanks!

**Review by macraig — APPROVED** [review]
LGTM!

**Review by malomarrec — APPROVED** [review]

**Review by lguychard — APPROVED** [review]
Really like this 🙂

File: content/direction/code-graph/index.md [hunk 1]
```diff
@@ -20,6 +20,25 @@ We build the Code Graph to make working with code easier, regardless of how much
...
 > We want to make using Sourcegraph to understand and modify code so easy that anyone can do it, including people not traditionally considered developers. We want to [democratize coding](../../company/strategy.md#10-year-vision) by making it so that everyone in the company can engage with the organization's code, safely making changes to implement their department's priorities and interests, without need for an intermediary to do it for them.
 
+## Principles
+
+### Abstraction vs. complexity
+
+There are two ways to think about categories of developer tools or features today:
+
+- Abstraction tools, which simplify tasks away by providing building blocks, sane defaults, and frameworks to build on. Most existing developer tools are of this variety.
+- Complexity-exploring tools, which help with complex problems such as finding and fixing issues in an existing, heterogeneous software system.. one that was created using the existing tools developers use today, with different languages, frameworks, and versions in constant flux, all interacting with each other.
```
Comment (benvenker) [inline content/direction/code-graph/index.md]: typo in "system.. one"?

**Review by benvenker — APPROVED** [review]
Looks good to me. Think I found a small typo, but other than it's great!

File: content/direction/code-graph/index.md [hunk 2]
```diff
@@ -20,6 +20,25 @@ We build the Code Graph to make working with code easier, regardless of how much
...
+
+This emerging need for complexity-exploring tooling is one we see Sourcegraph playing an important role in. The reality is that software tech stacks today look more like a rainforest — with animals and plants co-existing, competing, living, dying, growing, interacting in unplanned ways — than like a planned garden, and we are mindfully designing our the Code Graph to let you embrace this complexity rather than hiding it, or pushing you towards a specific single solution.
+
+Practically, this means a few things as we think about how the Code Graph should work:
+
+1. We can find and index your code wherever you host it or whatever language it is written in because we fit into your existing ecosystem.
```
Comment (Joelkw) [inline content/direction/code-graph/index.md]:
```suggestion
1. Sourcegraph fits in to your existing ecosystem – wherever you host your code, whatever language it's in, or however many repos you use. 
```
I've heard from others that we used to explicitly and do still try to uphold "treat monorepo and multi-repo codebases the same." This gets tricky (sometimes extremely tricky, in the case of code insights), but it's something that makes sense to me. If we agree and want to keep this going forward, I'd love to document that here somehow a bit more directly. 

I might also suggest changing "find and index" to be a bit broader, (is batch changes find but also "changing/acting/replacing"? Is code insights also "tracking"?) but don't have as strong a preference to that point

File: content/direction/code-graph/index.md [hunk 3]
```diff
@@ -20,6 +20,25 @@ We build the Code Graph to make working with code easier, regardless of how much
...
+1. By surfacing relationships (not just code itself, but data types, configution, and more) we help you build a clear, meaningful model of your software.
+1. We make this information easy to explore, with the most relevant information up front, easy to understand, and easy to share by revealing complexity as it is needed.
```
Comment (Joelkw) [inline content/direction/code-graph/index.md]:
```suggestion
1. We make this information easy to explore by revealing complexity as needed, with the most relevant information up front, easy to understand, and easy to share. 
```
(just fixing misplaced modifier)

**Review by Joelkw — APPROVED** [review]
glad to have this!

File: content/direction/code-graph/index.md (hunk 1 above)
Comment (jasonyavorsky) [reply content/direction/code-graph/index.md]:
```suggestion
- Complexity-exploring tools, which help with complex problems such as finding and fixing issues in existing, heterogeneous software systems. I.e., ones that were created using the existing tools developers use today, with different languages, frameworks, and versions in constant flux, all interacting with each other.
```

File: content/direction/code-graph/index.md (hunk 2 above)
Comment (jasonyavorsky) [reply content/direction/code-graph/index.md]: Really great improvement, thanks @Joelkw

**Pull request merged**